### PR TITLE
[fix #529] improved css font size validation

### LINKED ISSUE #529: Incorrect unit not detected
When validating epubcheck ignores an incorrect CSS syntax. 

I've got "1m" instead of "1em" and the epub validated with no problems or warnings (it crashed when opening with Adobe Digital Editions).

### PR DESCRIPTION
**Pull request by tofi86**
fixes #529

### CODE REVIEW

**Comment by tofi86** [comment]
I created a new testcase with db46daa which tests all supported property values and some invalid.

Test results BEFORE my changes, with 4.0.2:

```
→ epubcheck --mode exp -u /Users/tofi/Entwicklung/eclipse-workspace/default/epubcheck/src/test/resources/30/expanded/invalid/invalid-css-font-size-value
Validating using EPUB version 3.0.1 rules.
USAGE(ACC-007): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/content_001.xhtml(-1,-1): Content Documents do not use 'epub:type' attributes for semantic inflection.
USAGE(HTM-044): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/content_001.xhtml(1,100): Namespace uri 'http://www.idpf.org/2007/ops' was included but not used.
USAGE(ACC-008): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/invalid-css-font-size-value.epub(-1,-1): Navigation Document has no 'landmarks nav' element.
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(7,5): Value of CSS property 'font-size' does not use a relative size.
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(8,5): Value of CSS property 'font-size' does not use a relative size.
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(9,5): Value of CSS property 'font-size' does not use a relative size.
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(25,5): Value of CSS property 'font-size' does not use a relative size.
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(26,5): Value of CSS property 'font-size' does not use a relative size.
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(27,5): Value of CSS property 'font-size' does not use a relative size.
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(28,5): Value of CSS property 'font-size' does not use a relative size.
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(29,5): Value of CSS property 'font-size' does not use a relative size.
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(30,5): Value of CSS property 'font-size' does not use a relative size.
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(31,5): Value of CSS property 'font-size' does not use a relative size.
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(36,5): Value of CSS property 'font-size' does not use a relative size.
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(37,5): Value of CSS property 'font-size' does not use a relative size.
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(38,5): Value of CSS property 'font-size' does not use a relative size.
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(39,5): Value of CSS property 'font-size' does not use a relative size.
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(40,5): Value of CSS property 'font-size' does not use a relative size.
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(41,5): Value of CSS property 'font-size' does not use a relative size.
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(42,5): Value of CSS property 'font-size' does not use a relative size.
USAGE(HTM-021): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/content_001.xhtml(-1,-1): Content file doesn't contain lang attribute.
USAGE(HTM-021): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/nav.xhtml(-1,-1): Content file doesn't contain lang attribute.
Das EPUB enthält keine Fehler oder Warnungen.
Es ist valide.
EpubCheck abgeschlossen
```

and AFTER my changes with this branch:

```
Validating using EPUB version 3.0.1 rules.
USAGE(ACC-007): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/content_001.xhtml(-1,-1): Content Documents do not use 'epub:type' attributes for semantic inflection.
USAGE(ACC-008): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/nav.xhtml(-1,-1): Navigation Document has no 'landmarks nav' element.
USAGE(HTM-044): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/content_001.xhtml(1,100): Namespace uri 'http://www.idpf.org/2007/ops' was included but not used.
ERROR(CSS-020): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(7,5): CSS font selector declaration uses unexpected font-size value '1m'.
ERROR(CSS-020): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(8,5): CSS font selector declaration uses unexpected font-size value 'meter'.
ERROR(CSS-020): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(9,5): CSS font selector declaration uses unexpected font-size value 'in'.
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(26,5): Value of CSS property 'font-size' does not use a relative size: '1in'
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(27,5): Value of CSS property 'font-size' does not use a relative size: '1cm'
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(28,5): Value of CSS property 'font-size' does not use a relative size: '1mm'
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(29,5): Value of CSS property 'font-size' does not use a relative size: '1pt'
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(30,5): Value of CSS property 'font-size' does not use a relative size: '1pc'
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(31,5): Value of CSS property 'font-size' does not use a relative size: '1px'
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(36,5): Value of CSS property 'font-size' does not use a relative size: 'xx-small'
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(37,5): Value of CSS property 'font-size' does not use a relative size: 'x-small'
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(38,5): Value of CSS property 'font-size' does not use a relative size: 'small'
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(39,5): Value of CSS property 'font-size' does not use a relative size: 'medium'
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(40,5): Value of CSS property 'font-size' does not use a relative size: 'large'
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(41,5): Value of CSS property 'font-size' does not use a relative size: 'x-large'
USAGE(ACC-014): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/stylesheet.css(42,5): Value of CSS property 'font-size' does not use a relative size: 'xx-large'
USAGE(HTM-021): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/content_001.xhtml(-1,-1): Content file doesn't contain lang attribute.
USAGE(HTM-021): ./src/test/resources/30/expanded/invalid/invalid-css-font-size-value.epub/OPS/nav.xhtml(-1,-1): Content file doesn't contain lang attribute.

Check finished with errors

epubcheck completed
```

**Comment by tofi86** [comment]
@rdeltour I cannot reproduce the travis error ([link](https://travis-ci.org/IDPF/epubcheck/builds/249785151) for f4933dc) on my system:

```
MessageId_Coverage_Test(com.adobe.epubcheck.test.message_coverage)  Time elapsed: 0.016 sec  <<< FAILURE!

junit.framework.AssertionFailedError: Messages not covered by tests expected:
<[CHK-006, CHK-007, HTM-002, HTM-011, NAV-001, OPF-010, OPF-011, PKG-005, PKG-015, PKG-023, RSC-022]>
but was:
<[ACC-016, ACC-017, CHK-006, CHK-007, CSS-027, HTM-002, HTM-011, NAV-001, OPF-010, OPF-011, PKG-005, PKG-015, PKG-023, RSC-022]>
```

But I think I fixed it with 2368007

However, I have no Idea why those message ID's aren't covered anymore, as I didn't remove any of those tests.

**Comment by tofi86** [comment]
@rdeltour I really need some feedback from you on this topic in the next few weeks. Any help greatly appreciated. 🙏

File: src/test/resources/30/expanded/invalid/lorem-css-urls-2/EPUB/lorem.css [hunk 1]
```diff
@@ -1,16 +0,0 @@
-@charset 'utf-8';
-
-.trips_css_017_etc {
- font-size: 20;
```
Comment (rdeltour) [inline src/test/resources/30/expanded/invalid/lorem-css-urls-2/EPUB/lorem.css]: This used to trigger ACC_016

File: src/test/resources/30/expanded/invalid/lorem-css-urls-2/EPUB/lorem.css [hunk 2]
```diff
@@ -1,16 +0,0 @@
...
- line-height: 3.5px;
```
Comment (rdeltour) [inline src/test/resources/30/expanded/invalid/lorem-css-urls-2/EPUB/lorem.css]: This used to trigger ACC_017

**Comment by tofi86** [comment]
Thanks for the review! 

> That's actually a very good example of the maintenance cost of hidden side effects of not-enough-specific tests (in this case, the CSS triggered side effect usage warnings). I spent too much time on this.

Yeah, I'm totally with you, now that I manually diffed hundreds of JSON tests for PR #798  😞 

I will add additional tests for ACC_016 and ACC_017.

**Comment by tofi86** [comment]
> I will add additional tests for ACC_016 and ACC_017.

Well, so in the end I just reverted some unnecessary cleanup which I did in the middle of working at this PR.

Hopefully this will solve the coverage error and we can go ahead an merge this.

**Pull request merged**